MTM-41872: Define Codec API definition for implement a codec ms.

atulpandasag opened the pull request:
Initial commit for the LPWAN SDK which includes the API class and interface definitions.

atulpandasag committed:
Initial implementation. code formatting, clean up and few changes required.

atulpandasag committed:
Merge branch 'develop' of https://github.com/SoftwareAG/cumulocity-clients-java into develop

atulpandasag committed:
Merge branch 'develop' of https://github.com/SoftwareAG/cumulocity-clients-java into develop

atulpandasag committed:
MTM-41872: Initial commit for the LPWAN SDK

atulpandasag committed:
Code Formatting.

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/CodecMicroservice.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Set;
+
+/**
+ * This abstract class will be implemented to provide a set of unique supported devices.
```
<h1>Codec Microservice</h1>
CodecMicroservice should be implemented by the microservice implementer to advertise the devices it supports.

bbyreddy reviewed (commented):
Add copyright text to all the files, including the POM file.

bbyreddy reviewed (commented):
1. I thought we decided to remove the literal SDK from the package name.
2. Move the CodecMicroserviceApplication annotation into the annotation package.

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/CodecMicroserviceApplication.java:
```diff
@@ -0,0 +1,18 @@
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.TYPE})
```
- [ ] Add a comment to explain the purpose of this annotation and where the microservice implementer should be using it.

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/Decoder.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.stereotype.Component;
+
+/**
+ * This interface will be implemented to provide a decoding logic.
```
- [ ] <h1>Decode data received from a device</h1>
Decoder should be implemented to provide the decoding logic.

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/Decoder.java:
```diff
@@ -0,0 +1,24 @@
+@Component
+public interface Decoder {
+    /**
+     * This method should take the following params and should return the decoded response with mapping.
```
- [ ] This method should implement the logic to decode the data received from the device.
- [ ] describe the parameters payload, deviceId, deviceInfo and return class name and description.

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/DecodeResponse.java:
```diff
@@ -0,0 +1,17 @@
+    private List<EventRepresentation> eventsToCreate;
+    private List<AlarmRepresentation> alarmsToCreate;
+    private List<String> alarmTypesToClear;
+    private List<ManagedObjectProperty> propertiesToAdd;
```
- [ ] propertiesToAddToDeviceManagedObject

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/DeviceInfo.java:
```diff
@@ -0,0 +1,9 @@
+package com.cumulocity.lpwan.codec.sdk.model;
```
Add Java doc comments

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/model/ManagedObjectProperty.java:
```diff
@@ -0,0 +1,22 @@
+package com.cumulocity.lpwan.codec.sdk.model;
```
- [ ] Java doc comment

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java:
```diff
@@ -0,0 +1,51 @@
+    private Decoder decoder;
+
+    /**
+     * This method should decode the payload received by a particular device.
```
- [ ] This method invokes the decode logic provided by the microservice implementer by implementing the Decoder interface.

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * This method should register a device type upon subscribing the codec microservice.
```
- [ ] This method creates a Device Type managed object for every device supported by this microservice. Skips the creation if the Device Type already exists.

bbyreddy commented on c8y-lora-sdk/src/main/java/com/cumulocity/lpwan/codec/sdk/service/CodecService.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * This method should delete a device type upon unsubscribing the codec microservice.
```
- [ ] This method deletes the Device Type managed object created earlier for every device supported by this microservice. Skips the deletion if the Device Type is associated with a device.

bbyreddy commented on c8y-lora-sdk/src/test/resources/application.properties:
```diff
@@ -0,0 +1,3 @@
+application.name=c8y-lora-sdk
+server.port=8071
```
- [ ] Is this server.port property mandated for every microservice?

atulpandasag committed:
1. Added Copyright text to all the files
2. Removed 'sdk' from the package name
3. Added a ne package called 'annotation'
4. Updated the java doc for CodecMicroservice.
5. Added Java doc for CodecMicroserviceApplication.

atulpandasag committed:
New Changes for Codec SDK
1. Register and Unregister device type
2. Changed input params for Decode request.

Commit:
Update Constants.java

Formatted Constants file.

atulpandasag committed:
Initial commit for calling the codec microservice

atulpandasag committed:
Changes fofr DecodePayload

atulpandasag committed:
new changes including names,models.

atulpandasag committed:
Invoke call to codec microservice
Executing MEAs

atulpandasag committed:
new changes.

atulpandasag committed:
new changes.

Commit:
Update ManagedObjectProperty.java

atulpandasag committed:
Changed the sdk name.
Added new method to DecoderOutput

atulpandasag committed:
Update device type if already exists.

atulpandasag committed:
Updated External Id of teh devicetype to contain the fieldbus type

atulpandasag committed:
Validate DeviceInfo and DecoderInput for null entries.

atulpandasag committed:
Removed Network provider enum.

atulpandasag committed:
Updated DeviceInfo instance creation.
Removed commented code.

Commit:
Removed getDeviceTypeName method from DeviceInfo class and removed unused constants and imports.

Commit:
Updated the snapshot version.

Commit:
Input validation and proper exception handling.

Commit:
Added unit tests, with coverage of 94% lines.

atulpandasag committed:
Updated test cases.

atulpandasag committed:
Unit test for PayloadDecoderService

atulpandasag committed:
Updated test cases.

atulpandasag committed:
Reverted code.

atulpandasag committed:
Optimized imports

Commit:
Updated the ExceptionsHandler to handle throwable instead of a specific exception class.

Commit:
Added @NotNull and @NotBlank for the method parameters accordingly.

atulpandasag committed:
Added Unit Test and a new annotation for webclient bean

atulpandasag committed:
Updated Webclient implementation and updated the relevant test cases.

atulpandasag committed:
Webclient update

atulpandasag committed:
Added Codec.java
Moved Decoder to a new package.

Commit:
Added some convenient methods to DecoderInput and DecoderOutput classes.

Commit:
Adopted the ManagedObjectProperty changes in LPWAN backend library.

Commit:
Some more refactoring and also changed the deviceManagedObjectToUpdate to a singular object from a list.

Commit:
Renamed the codec sdk folder, artifactId and groupId to device-codec-sdk.

Commit:
Updated the dependency to the renamed device-codec-sdk module.

Commit:
Changed the root package name to replace the reference to lpwan.codec with device.codec.

Commit:
Renamed the folder and module to lpwan-codec-sdk

Commit:
Some more changes to the convenience methods in DecoderOutput

Commit:
Fixed issues in DecoderOutput and added some more tests.

Commit:
Moved DecoderException class into the decoder package.

Commit:
Cosmetic changes

atulpandasag committed:
Included Javadoc

Commit:
Minor changes so that the parameter validation is done first.

Commit:
Updated maven file.

atulpandasag committed:
Updated Javadoc

Commit:
Squashed commit of the following:

commit 2684c5d9459f8544cc7cc0045189da31b8347a4b
Merge: 33939c22d 70eabdec5
Author: Bhaskar Reddy Byreddy <bbyreddy@softwareag.com>
Date:   Tue Dec 7 05:43:58 2021 +0000

    Merge branch 'MTM-41872-define-codec-api-definition-for-implement-a-codec-ms' into MTM-41762-lpwan-codec-based-on-microservice-custom-decoders-

commit 33939c22d7c883ac56f0c92f2b02e74feb69945f
Author: Bhaskar Reddy Byreddy <bbyreddy@softwareag.com>
Date:   Tue Dec 7 05:33:22 2021 +0000

    LPWAN device specific custom codec framework based on the microservice/custom-decoders API.

commit 1b019903b5be5f96e6e37b21df1c1de6ccb6e712
Author: Bhaskar Reddy Byreddy <bbyreddy@softwareag.com>
Date:   Mon Dec 6 17:06:11 2021 +0000

    LPWAN device specific custom codec framework based on the microservice/custom-decoders API.

commit 385330aa58bd3ef926de616cccbb7f11912fc4ee
Author: Bhaskar Reddy Byreddy <bbyreddy@softwareag.com>
Date:   Mon Dec 6 17:00:59 2021 +0000

    LPWAN device specific custom codec framework based on the microservice/custom-decoders API.

commit 36610aa1bf00ceb6457347b3eb2655542bc255a2
Author: Bhaskar Reddy Byreddy <bbyreddy@softwareag.com>
Date:   Mon Dec 6 16:57:04 2021 +0000

    Added alarmTypesToClear to the DecoderResult class.
    Added capability to handle a non string value to the DataFragmentUpdate

Commit:
Modified the tests.

Commit:
Minor changes to the maven file.

Commit:
Modified the tests.

Commit:
Modified the tests.

atulpandasag committed:
Added AlarmTypes to the list of alarm types to be cleared.

atulpandasag committed:
Fixed Unit tests for PayloadMappingService#handleCodecServiceResponse()

atulpandasag committed:
Updated Javadoc

Commit:
Updated the lpwan specific keys to be passed to /decode service under args parameter.

atulpandasag committed:
Getting context path via application api.

atulpandasag committed:
Removed Unused projects.
Updated Test cases related to context path

atulpandasag committed:
Merge branch 'develop' of https://github.com/SoftwareAG/cumulocity-clients-java into develop

atulpandasag committed:
Updated pom file

atulpandasag committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

# Conflicts:
#	lpwan-backend/pom.xml
#	lpwan-backend/src/main/java/com/cumulocity/lpwan/payload/service/PayloadDecoderService.java

atulpandasag committed:
Added Unit Test and a new annotation for webclient bean

atulpandasag committed:
Removed unused code introduced while resolving rebase and merge conflict

atulpandasag committed:
Added Alarm types to update.

atulpandasag committed:
Added Alarm Type to update parsing

atulpandasag committed:
Added Unit test cases for Update Alarm Types.

atulpandasag committed:
Added External id info of the device type to the device managed object.

Commit:
LPWAN Encoding Code

Commit:
Encoding unit tests and remove unused imports

Commit:
No Args constructor for deserialization purpose and annotate with JSONProperty as we are passing deviceManufacturer and deviceModel in input

Commit:
Utility functions and tests for them

Commit:
Javadoc for encoding functions, Deserialization NoArgs Constructor fix, Added PredefinedCommandFilter InventoryFilter derived class as the existing implementation was containing the CodecMicroservice object causing ClassCastException and other misc changes

miryalarahul commented on microservice/custom-encoders/pom.xml:
```diff
@@ -0,0 +1,27 @@
+    <parent>
+        <artifactId>clients-java</artifactId>
+        <groupId>com.nsn.cumulocity.clients-java</groupId>
+        <version>1011.67.0-SNAPSHOT</version>
```
please remove this hard coded version and maintain like other pom version

miryalarahul commented on microservice/pom.xml:
```diff
@@ -38,10 +38,12 @@
         <module>subscription</module>
         <module>logging</module>
+        <module>custom-decoders</module>
+        <module>custom-encoders</module>
```
check the order is correct or not

Commit:
Rename manufacturer and model to deviceManufacturer and deviceModel and kept constants only in Device Info

atulpandasag committed:
Merge remote-tracking branch 'origin/develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

atulpandasag committed:
Merge branch 'MTM-41872-define-codec-api-definition-for-implement-a-codec-ms' of https://github.com/SoftwareAG/cumulocity-clients-java into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

atulpandasag committed:
Merge remote-tracking branch 'origin/develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

atulpandasag committed:
Updated POM Version.

Commit:
Test Code fix compilation errors

Commit:
Merge remote-tracking branch 'origin/MTM-41872-define-codec-api-definition-for-implement-a-codec-ms' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

Commit:
Test Code fix compilation errors 2

atulpandasag committed:
Merge remote-tracking branch 'origin/develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

Commit:
Add copyright to files

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

Commit:
Readme file for custom microservice

Commit:
Javadoc build errors fix

Commit:
Javadoc build errors fix 2

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

Commit:
Refactor EncoderInputData

Commit:
Merge remote-tracking branch 'origin/MTM-41872-define-codec-api-definition-for-implement-a-codec-ms' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

atulpandasag committed:
Corrected the baseUrl.

Commit:
Use DecoderInputData instead of LpwanDecoderInputData and readme change

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

Commit:
Readme change

Commit:
Merge remote-tracking branch 'origin/MTM-41872-define-codec-api-definition-for-implement-a-codec-ms' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

atulpandasag committed:
Merge remote-tracking branch 'origin/develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

ashhSAG93 committed:
Merge branch 'develop' into MTM-41872-define-codec-api-definition-for-implement-a-codec-ms

deepakgoenka commented:
This PR needs to be reverted as post-merge has got into a bad shape due to few recent PRs
